### ISSUE TITLE
Container only listens ipv6 ports

### ISSUE THREAD

**Issue by KrisTheGreat** [bgauduch/call-of-duty-2-docker-server#21]
Hi,
I have a problem with a container networking. The container is listening only for ipv6 connections. I have some more containers which doesnt't have this behavior na run fine at ipv4.

```
netstat -ntulp
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 0.0.0.0:25              0.0.0.0:*               LISTEN      1270/master
tcp        0      0 0.0.0.0:443             0.0.0.0:*               LISTEN      1038/nginx: master
tcp        0      0 127.0.0.1:4000          0.0.0.0:*               LISTEN      1881/docker-proxy-c
tcp        0      0 127.0.0.1:5000          0.0.0.0:*               LISTEN      3266/docker-proxy-c
tcp        0      0 127.0.0.1:27017         0.0.0.0:*               LISTEN      3282/docker-proxy-c
tcp        0      0 127.0.0.1:2000          0.0.0.0:*               LISTEN      1896/docker-proxy-c
tcp        0      0 0.0.0.0:80              0.0.0.0:*               LISTEN      1038/nginx: master
tcp        0      0 127.0.0.1:8081          0.0.0.0:*               LISTEN      3238/docker-proxy-c
tcp        0      0 127.0.0.1:1234          0.0.0.0:*               LISTEN      3551/docker-proxy-c
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      1315/sshd
tcp6       0      0 :::28960                :::*                    LISTEN      3683/docker-proxy-c
tcp6       0      0 :::22                   :::*                    LISTEN      1315/sshd
udp        0      0 0.0.0.0:68              0.0.0.0:*                           825/dhclient
udp        0      0 127.0.0.1:323           0.0.0.0:*                           607/chronyd
udp6       0      0 :::20500                :::*                                3716/docker-proxy-c
udp6       0      0 :::20510                :::*                                3705/docker-proxy-c
udp6       0      0 :::28960                :::*                                3694/docker-proxy-c
udp6       0      0 ::1:323                 :::*                                607/chronyd

````

**Comment by bgauduch** [comment]
Hey there @KrisTheGreat !

Thanks for opening this issue, I will investigate right after I've fixed the build to merge branch #20 !

**Comment by bgauduch** [comment]
Hey @KrisTheGreat, I cannot reproduce this issue locally, here is the output of netstat I get:
```bash
$ netstat -ntl
Active Internet connections
Proto Recv-Q Send-Q  Local Address          Foreign Address        (state) 
udp46      0      0  *.28960                *.*                               
udp46      0      0  *.20510                *.*                               
udp46      0      0  *.20500                *.* 
[REDACTED]
````
As we can see, ports are exposed both for ipv4 and ipv6 protocol.

What is your output for `docker container ls`? Just to be sure docker is doing the correct port forwarding and mapping to the host.

Mine give this:
```bash
$ docker container ls
CONTAINER ID        IMAGE                      COMMAND             CREATED             STATUS              PORTS                                                                                                    NAMES
79e37d7422a2        bgauduch/cod2-server:dev   "/entrypoint.sh"    18 minutes ago      Up 18 minutes       0.0.0.0:20500->20500/udp, 0.0.0.0:20510->20510/udp, 0.0.0.0:28960->28960/udp, 0.0.0.0:28960->28960/tcp   call-of-duty-2-docker-server_cod2_server_1
```

I've found an issue that might help you: https://github.com/moby/moby/issues/2174#issuecomment-35219757

But I believe this issue might be related to the way your docker daemon  and / or network interfaces are configured.

**Comment by bgauduch** [comment]
Hello @KrisTheGreat, any update on this issue ?

Did you did you resolve it ?

**Comment by bgauduch** [comment]
Hi @KrisTheGreat, without hearing back from you I will close this issue.

Feel free to re-open it if needed !

**Comment by Betalord** [comment]
I think I'm having the same issue. I can't get the server up, the ports are closed. It seems ipv6 ports are open, but not v4. Here is my log:

```
netstat -ntulp
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 127.0.0.1:38127         0.0.0.0:*               LISTEN      6837/containerd
tcp        0      0 0.0.0.0:80              0.0.0.0:*               LISTEN      786643/apache2
tcp        0      0 0.0.0.0:10000           0.0.0.0:*               LISTEN      587/perl
tcp        0      0 127.0.0.53:53           0.0.0.0:*               LISTEN      416/systemd-resolve
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      480/sshd: /usr/sbin
tcp        0      0 0.0.0.0:443             0.0.0.0:*               LISTEN      786643/apache2
tcp        0      0 0.0.0.0:3306            0.0.0.0:*               LISTEN      5012/mysqld
tcp        0      0 127.0.0.1:6379          0.0.0.0:*               LISTEN      787656/redis-server
tcp6       0      0 :::22000                :::*                    LISTEN      581075/syncthing
tcp6       0      0 :::22                   :::*                    LISTEN      480/sshd: /usr/sbin
tcp6       0      0 :::28960                :::*                    LISTEN      791342/docker-proxy
tcp6       0      0 :::8384                 :::*                    LISTEN      581075/syncthing
tcp6       0      0 :::33060                :::*                    LISTEN      5012/mysqld
udp        0      0 127.0.0.53:53           0.0.0.0:*                           416/systemd-resolve
udp        0      0 0.0.0.0:21027           0.0.0.0:*                           581075/syncthing
udp        0      0 0.0.0.0:10000           0.0.0.0:*                           587/perl
udp        0      0 0.0.0.0:34600           0.0.0.0:*                           581075/syncthing
udp6       0      0 :::20500                :::*                                791381/docker-proxy
udp6       0      0 :::20510                :::*                                791368/docker-proxy
udp6       0      0 :::28960                :::*                                791354/docker-proxy
udp6       0      0 :::21027                :::*                                581075/syncthing
udp6       0      0 :::41894                :::*                                581075/syncthing
udp6       0      0 :::22000                :::*                                581075/syncthing
```

Here is the log from the game server (not sure if you need it though):

```
cod2_server_1  | -----------------------------------
cod2_server_1  | Resolving cod2master.activision.com
cod2_server_1  | cod2master.activision.com resolved to 185.34.107.159:20710
cod2_server_1  | Hitch warning: 4390 msec frame time
cod2_server_1  | Resolving cod2master.activision.com
cod2_server_1  | cod2master.activision.com resolved to 185.34.107.159:20710
cod2_server_1  | Sending heartbeat to cod2master.activision.com
cod2_server_1  | Resolving master.cod2.ru
cod2_server_1  | master.cod2.ru resolved to 185.211.247.23:20710
cod2_server_1  | Sending heartbeat to master.cod2.ru
cod2_server_1  | Sending heartbeat to cod2master.activision.com
cod2_server_1  | Sending heartbeat to master.cod2.ru
cod2_server_1  | Sending heartbeat to cod2master.activision.com
cod2_server_1  | Sending heartbeat to master.cod2.ru
cod2_server_1  | Sending heartbeat to cod2master.activision.com
cod2_server_1  | Sending heartbeat to master.cod2.ru

```

**Comment by bgauduch** [comment]
Hello @Betalord !

As stated in my [previous answer](https://github.com/bgauduch/call-of-duty-2-docker-server/issues/21#issuecomment-605136046), I don't think this is the root cause.

See [here](https://github.com/moby/moby/issues/2174#issuecomment-31701780) for explanations.

Can you provide me the complete server logs output, so that I can check that it started correctly ?

**Comment by Betalord** [comment]
Sure thing, thanks for your help! Here is the log:

```
root@vmi378274:/home/betalord/cod2/call-of-duty-2-docker-server# ./scripts/dev-logs.sh
Attaching to cod2
cod2           | > [LIBCOD] Compiled for: CoD2 1.3
cod2           | > [LIBCOD] Compiled May  4 2020 07:40:17 using GCC 8.3.0
cod2           | > [PLUGIN LOADED]
cod2           | CoD2 MP 1.3 build linux-i386 Jun 23 2006
cod2           | ----- FS_Startup -----
cod2           | Current language: english
cod2           | Current search path:
cod2           | /home/cod2/.callofduty2/main
cod2           | /home/cod2/main/iw_03.iwd (74 files)
cod2           | /home/cod2/main/iw_02.iwd (3 files)
cod2           | /home/cod2/main/iw_01.iwd (5174 files)
cod2           | /home/cod2/main/iw_00.iwd (20214 files)
cod2           | /home/cod2/main
cod2           | /home/cod2/.callofduty2/raw
cod2           | /home/cod2/.callofduty2/raw_shared
cod2           | /home/cod2/.callofduty2/devraw
cod2           | /home/cod2/.callofduty2/devraw_shared
cod2           | /home/cod2/raw
cod2           | /home/cod2/raw_shared
cod2           | /home/cod2/devraw
cod2           | /home/cod2/devraw_shared
cod2           | /home/cod2/main/localized_english_iw02.iwd (6 files)
cod2           |     localized assets iwd file for english
cod2           | /home/cod2/main/localized_english_iw01.iwd (15156 files)
cod2           |     localized assets iwd file for english
cod2           | /home/cod2/main/localized_english_iw00.iwd (287 files)
cod2           |     localized assets iwd file for english
cod2           |
cod2           | File Handles:
cod2           | ----------------------
cod2           | 40914 files in iwd files
cod2           | execing default_mp.cfg
cod2           | couldn't exec language.cfg
cod2           | couldn't exec config_mp_server.cfg
cod2           | dedicated is read only.
cod2           | Opening IP socket: localhost:28960
cod2           | Hostname: d4af5b45819d
cod2           | IP: 172.18.0.2
cod2           | --- Common Initialization Complete ---
cod2           | dedicated is read only.
cod2           | net_port is write protected.
cod2           | sv_punkbuster is write protected.
cod2           | execing punkbuster.cfg
cod2           | execing server_mp.cfg
cod2           | logfile opened on Mon May 11 09:27:27 2020
cod2           |
cod2           | map_rotate...
cod2           |
cod2           | "sv_mapRotation" is:"gametype tdm map mp_brecourt gametype ctf map mp_carentan gametype tdm map mp_farmhouse gametype ctf map mp_dawnville gametype tdm map mp_burgundy gametype ctf map mp_trainstation gametype tdm map mp_decoy gametype ctf map mp_toujane gametype tdm map mp_matmata gametype ctf map mp_railyard gametype tdm map mp_breakout gametype ctf map mp_leningrad gametype tdm map mp_downtown"
cod2           |
cod2           | "sv_mapRotationCurrent" is:""
cod2           |
cod2           | Setting g_gametype: tdm.
cod2           | Setting map: mp_brecourt.
cod2           | ------ Server Initialization ------
cod2           | Server: mp_brecourt
cod2           | ----- FS_Startup -----
cod2           | Current language: english
cod2           | Current search path:
cod2           | /home/cod2/.callofduty2/main
cod2           | /home/cod2/main/iw_03.iwd (74 files)
cod2           | /home/cod2/main/iw_02.iwd (3 files)
cod2           | /home/cod2/main/iw_01.iwd (5174 files)
cod2           | /home/cod2/main/iw_00.iwd (20214 files)
cod2           | /home/cod2/main
cod2           | /home/cod2/.callofduty2/raw
cod2           | /home/cod2/.callofduty2/raw_shared
cod2           | /home/cod2/.callofduty2/devraw
cod2           | /home/cod2/.callofduty2/devraw_shared
cod2           | /home/cod2/raw
cod2           | /home/cod2/raw_shared
cod2           | /home/cod2/devraw
cod2           | /home/cod2/devraw_shared
cod2           | /home/cod2/main/localized_english_iw02.iwd (6 files)
cod2           |     localized assets iwd file for english
cod2           | /home/cod2/main/localized_english_iw01.iwd (15156 files)
cod2           |     localized assets iwd file for english
cod2           | /home/cod2/main/localized_english_iw00.iwd (287 files)
cod2           |     localized assets iwd file for english
cod2           |
cod2           | File Handles:
cod2           | handle 1: console_mp_server.log
cod2           | ----------------------
cod2           | 40914 files in iwd files
cod2           | ------- Game Initialization -------
cod2           | gamename: Call of Duty 2
cod2           | gamedate: Jun 23 2006
cod2           |   0:00 ------------------------------------------------------------
cod2           |   0:00 InitGame: \_Admin\Betalord\_Email\not.given@gmail.com\_Location\YourCountry\_Website\betalord.com\g_antilag\0\g_gametype\tdm\gamename\Call of Duty 2\mapname\mp_brecourt\protocol\118\shortversion\1.3\sv_allowAnonymous\0\sv_floodProtect\1\sv_hostname\Betalord's server\sv_maxclients\20\sv_maxPing\150\sv_maxRate\25000\sv_minPing\0\sv_privateClients\0\sv_punkbuster\0\sv_pure\0\sv_voice\1
cod2           | -----------------------------------
cod2           | -----------------------------------
cod2           | Resolving cod2master.activision.com
cod2           | cod2master.activision.com resolved to 185.34.107.159:20710
cod2           | Hitch warning: 4236 msec frame time
cod2           | Resolving cod2master.activision.com
cod2           | cod2master.activision.com resolved to 185.34.107.159:20710
cod2           | Sending heartbeat to cod2master.activision.com
cod2           | Resolving master.cod2.ru
cod2           | master.cod2.ru resolved to 185.211.247.23:20710
cod2           | Sending heartbeat to master.cod2.ru
cod2           | status
cod2           | map: mp_brecourt
cod2           | num score ping guid   name            lastmsg address               qport rate
cod2           | --- ----- ---- ------ --------------- ------- --------------------- ----- -----
cod2           |
cod2           | Sending heartbeat to cod2master.activision.com
cod2           | Sending heartbeat to master.cod2.ru

```

Here is the procedure I used to set up the container:

```
docker pull bgauduch/cod2server:latest
// Then I copied over cod2 main data files, modified the config, etc.

docker-compose up -d
docker rename call-of-duty-2-docker-server_cod2_server_1 cod2
root@vmi378274:/home/betalord/cod2/call-of-duty-2-docker-server# docker container ls -a
CONTAINER ID        IMAGE                        COMMAND                  CREATED             STATUS              PORTS                                                                                                    NAMES
d4af5b45819d        bgauduch/cod2server:latest   "./cod2_lnxded +set …"   9 minutes ago       Up 9 minutes        0.0.0.0:20500->20500/udp, 0.0.0.0:20510->20510/udp, 0.0.0.0:28960->28960/udp, 0.0.0.0:28960->28960/tcp   cod2
```

**Comment by Betalord** [comment]
Also, this is the result of online port scan for ipv4:
[scrshot ipv4](https://betalord.com/tmp/ipv4_scan.png)

And for ipv6:
[scrshot ipv6](https://betalord.com/tmp/ipv6_scan.png)
